pwnlib/fmtstr.py

```diff
@@ -102,7 +108,497 @@ def send_payload(payload):
+        return self.__key() == other.__key()
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash(self.__key())
+
+    def __repr__(self):
+        return "AtomWrite(start=%d, size=%d, integer=%#x, mask=%#x)" % (self.start, self.size, self.integer, self.mask)
+
+    @property
+    def bitsize(self):
+        return self.size * 8
+
+    @property
+    def end(self):
+        return self.start + self.size
+
+    def compute_padding(self, counter):
+        """
+        This function computes the least amount of padding necessary to execute this write,
+        given the current format string write counter (how many bytes have been written until now).
+
+        Examples:
+        >>> hex(pwnlib.fmtstr.AtomWrite(0x0, 0x2, 0x2345).compute_padding(0x1111))
+        0x1234
+        >>> hex(pwnlib.fmtstr.AtomWrite(0x0, 0x2, 0xaa00).compute_padding(0xaabb))
+        0xff45
+        >>> hex(pwnlib.fmtstr.AtomWrite(0x0, 0x2, 0xaa00, 0xff00).compute_padding(0xaabb)) # with mask
+        0x0
+        """
+        wanted = self.integer & self.mask
+        padding = 0
+        while True:
+            diff = wanted ^ ((counter + padding) & self.mask)
+            if not diff: break
+            # this masks the least significant set bit and adds it to padding
+            padding += diff & (diff ^ (diff - 1))
+        return padding
+
+    def replace(self, start=None, size=None, integer=None, mask=None):
+        """
+        Return a new write with updated fields (everything that is not None is set to the new value)
+        """
+        start = self.start if start is None else start
+        size = self.size if size is None else size
+        integer = self.integer if integer is None else integer
+        mask = self.mask if mask is None else mask
+        return AtomWrite(start, size, integer, mask)
+
+    def union(self, other):
+        """
+        Combine adjacent writes into a single write.
+
+        Example:
+        >>> context.clear(endian = "little")
+        >>> pwnlib.fmtstr.AtomWrite(0x0, 0x1, 0x1, 0xff).union(pwnlib.fmtstr.AtomWrite(0x1, 0x1, 0x2, 0x77))
+        AtomWrite(start=0, size=2, integer=0x201, mask=0xff77)
+        """
+        assert other.start == self.end, "writes to combine must be continous"
+        if context.endian == "little":
+            newinteger = (other.integer << self.bitsize) | self.integer
+            newmask = (other.mask << self.bitsize) | self.mask
+        elif context.endian == "big":
+            newinteger = (self.integer << other.bitsize) | other.integer
+            newmask = (self.mask << other.bitsize) | other.mask
+        return AtomWrite(self.start, self.size + other.size, newinteger, newmask)
+
+    def __getslice__(self, i,  j):
+        return self.__getitem__(slice(i, j))
+
+    def __getitem__(self, i):
+        if type(i) is not slice:
```

Comment:
Preferrable type checking is by `isinstance()`
```suggestion
        if not isinstance(i, slice):
```